setup.py

```diff
@@ -14,15 +17,26 @@
 with io_open(version_file, mode='r') as fd:
     exec(fd.read())
 
-install_requires = ['numpy',
-                    'imageio',
-                    'pillow',
-                    'appdirs',
-                    'scooby>=0.5.1',
-                    'vtk',
-                    'dataclasses;python_version=="3.6"',
-                    'typing_extensions;python_version<="3.7"',
-                    ]
+install_requires = [
+    'numpy',
+    'imageio',
+    'pillow',
+    'appdirs',
+    'scooby>=0.5.1',
+    'vtk',
+    'dataclasses;python_version=="3.6"',
+    'typing_extensions;python_version<="3.7"',
+]
+
+# Handle Python 3.10
+# remove this when Kitware releases VTK on PyPI as we prefer official wheels
+is64 = struct.calcsize("P") * 8 == 64
+if sys.version_info.minor == 10 and is64:
+    if platform.system().lower() == 'linux':
+        install_requires.remove('vtk')
+        install_requires.append(
+            'vtk @ https://github.com/pyvista/pyvista-wheels/raw/main/vtk-9.1.0.dev0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl',
```

Comment:
Does this implementation prevent users from compiling their own vtk wheel with a more up to date vtk release?

Is it better to give them an error with a copy-paste-install suggestion like ``A custom built wheel for a dev version of vtk-9.1.0 is required.  You can install (an out of date) version with `pip install ..` ``

Without this, it will be unclear to the user that they are using a nonstandard vtk installation.  My preference is to get explicit buy in from the user that they are doing something non-standard.

If you prefer the current method as there is a one-line install, I would put in a warning that an out of date dev version is being installed.

Reply:
I'm not actually sure how this works in the whole packaging process when users install pyvista, so maybe my thoughts don't make practical sense.

Reply:
@MatthewFlamm I started a new venv, installed official VTK wheels in it, then cloned pyvista, edited the setup.py hardcoding a similar requirement item using the pyvista wheels (for the relevant python version), and then running `pip install .` purged the original VTK and installed the pyvista ones.

Which is to say, indeed this seems like breaking pre-existing VTK wheels. I agree we shouldn't do this. A lot of people probably build their own VTK wheels (at least one person mentioned this in slack, exactly because they want to use the newest Python release). We shouldn't make this impossible. If some quirk/bug is in the dev version of VTK that we vendor, users will have no way of working around that, or even try debugging.

I'm aware this is "temporary" support, but I'd like for us to explore alternatives. Is there a way to detect during PyVista's install that vtk is not available in the given environment? Or catching build errors from within setup.py? (With reasonable amounts of black magic.) Even if not, simply letting pip install to fail without PyPI 3.10 VTK wheels might be my preference. We can try adding warnings/hints on PyPI and the website, installation instructions etc. (and even pin an issue on the repo) that explains that we have VTK wheels for 3.10 if anyone needs it.

Reply:
How about just:
```
try:
    import vtk
except:
    install_requires.append(...
```

Reply:
@akaszynski I like it!

One (probably very) edge case: someone tries to install their own wheels _and_ pyvista in the same pip call (assuming that is even supposed to work), in which case I'm not sure if pyvista will have access to the vtk. Perhaps it would. And in any case I'd expect self-built wheels to be installed separately. Unless there's some CI pipeline that could try doing this.

I don't have time right now to check what `pip install this_weird_vtk.whl pyvista` would do in this situation. It's very unlikely. I'm only musing about this because _if_ this happens, it would lead to successful install with a semi-silent swap of the expected VTK wheels with our own.

Reply:
Turns out there are a few edge cases. First and foremost: *.whl won't include logic from our `setup.py` and therefore we can't dynamically change our `install_requires`.  For example, consider the following in `setup.py`:

##### `setup.py`
```py
# unofficial wheels for python 3.10
WHEELS_URL = 'https://github.com/pyvista/pyvista-wheels/raw/main'
unoffical_wheels = {
    'linux': f'{WHEELS_URL}/vtk-9.1.0.dev0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl',
    'windows': f'{WHEELS_URL}/vtk-9.1.0.dev0-cp310-cp310-win_amd64.whl',
}

# Handle Python 3.10
# remove this when Kitware releases VTK on PyPI
is64 = struct.calcsize("P") * 8 == 64
if sys.version_info.minor == 10 and is64:
    # use pip to check if vtk is available or installed
    try:
        import pip
        # returns non zero when installation fails
        out = pip.main(['install', 'vtk'])
    except ImportError:
        # in the unlikely case where pip isn't installed
        out = 0

    if out:
        raise RuntimeError(f"out:{out}")
        sys_name = platform.system().lower()
        if sys_name in unoffical_wheels:
            raise RuntimeError(
                '\n\n\n'
                'VTK for Python 3.10 is not out yet on PyPI, but you can try '
                'installing unofficial wheels we publish for your platform with:\n\n'
                f'    pip install {unoffical_wheels[sys_name]}\n\n\n'
            )
```

This won't work when installing from a wheel. I think the best approach is:

>  We can try adding warnings/hints on PyPI and the website, installation instructions etc. (and even pin an issue on the repo) that explains that we have VTK wheels for 3.10 if anyone needs it.

Let's just pin an issue or discussion and point people in the right direction.

Reply:
Thanks for trying to make it work @akaszynski

Reply:
Thanks! I'm going to use custom vtk in the pinned repositories in https://github.com/pyansys/, but I agree that for this library, let's stick with the official wheels.

Reply:
FWIW, all of this can be avoided by using the package index I set up in https://github.com/pyvista/pyvista-wheels 's workflows to publish on GitHub pages at: https://pyvista.github.io/pyvista-wheels/ and https://wheels.pyvista.org/

To use this, simply add `--find-links` as a `pip install` option and it will give preference to our wheels there. So users who want to opt-in to using our wheels can do so by running:

```
pip install --find-links https://wheels.pyvista.org/ pyvista
```

This is still very much a work in progress, as https://github.com/pyvista/pyvista-wheels/issues/8 prevents this from working properly and I need to update it to use https://github.com/banesullivan/create-pip-index-action

----

I've been doing this a lot lately at work to ship custom builds of software or small Python packages we don't want on PyPI

----

Also, @akaszynski, I can probably get python 3.10 wheels of VTK uploaded to PyPI directly later this week (I just need to build them for all platforms)

Reply:
I added a macos wheel so at this point `pip install --find-links https://wheels.pyvista.org/ pyvista` should work on Python 3.10 for all platforms:

```
$ python
Python 3.10.2 | packaged by conda-forge | (main, Feb  1 2022, 19:32:15) [Clang 11.1.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
No jedi here. Coming to the dark side.
>>>

$ pip install --no-cache  --find-links https://wheels.pyvista.org pyvi
sta
Looking in links: https://wheels.pyvista.org
Collecting pyvista
  Downloading pyvista-0.33.2-py3-none-any.whl (1.4 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 24.1 MB/s eta 0:00:00
Collecting numpy
  Downloading numpy-1.22.2-cp310-cp310-macosx_10_14_x86_64.whl (17.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 78.9 MB/s eta 0:00:00
Collecting pillow
  Downloading Pillow-9.0.1-cp310-cp310-macosx_10_10_universal2.whl (3.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 66.0 MB/s eta 0:00:00
Collecting appdirs
  Downloading appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
Collecting scooby>=0.5.1
  Downloading scooby-0.5.11-py3-none-any.whl (14 kB)
Collecting vtk
  Downloading https://wheels.pyvista.org/vtk-9.1.0.dev0-cp310-cp310-macosx_12_0_x86_64.whl (38.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.0/38.0 MB 78.7 MB/s eta 0:00:00
Collecting imageio
  Downloading imageio-2.14.1-py3-none-any.whl (3.3 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 69.4 MB/s eta 0:00:00
Installing collected packages: vtk, scooby, appdirs, pillow, numpy, imageio, pyvista
Successfully installed appdirs-1.4.4 imageio-2.14.1 numpy-1.22.2 pillow-9.0.1 pyvista-0.33.2 scooby-0.5.11 vtk-9.1.0.dev0
```